shrink

```diff
@@ -1,82 +1,269 @@
 #!/bin/bash
-cd
+# Checked by https://www.shellcheck.net/
+
+
+# Constants
 MYSQL_USER=root
-MYSQL_PASS=root
-MYSQL_SERVICE=mysql
-MYSQL=$(which mysql)
-if [ $? == 1 ]; then
-  echo "MySQL is not installed."
-  exit 1
-fi
-MYSQL_DUMP=$(which mysqldump)
-if [ $? == 1 ]; then
-  echo "Unable to execute mysqldump."
-  exit 1
-fi
-if test $1; then
-  if [[ "$1" != "-k" ]]; then
-    echo "Invalid switch $1."
-    echo "Usage: $0 [-k]"
-    echo "Use the -k switch to keep backups instead of restoring the databases"
-    exit 1
-  fi
-fi
-MYSQLD=$(pidof mysqld)
-if [ -z "$MYSQLD" ]; then
-  sudo service $MYSQL_SERVICE start
-fi
-if ! $MYSQL -u$MYSQL_USER -p$MYSQL_PASS  -e ';' 2> /dev/null; then
-  read -sp "Enter the MySQL $MYSQL_USER password: " MYSQL_PASS
-  if ! $MYSQL -u$MYSQL_USER -p$MYSQL_PASS  -e ';' 2> /dev/null; then
-    echo "Invalid password."
-    exit 1
-  fi
-fi
-DBS="$($MYSQL -u$MYSQL_USER -p$MYSQL_PASS -e 'show databases' | awk '{print $1}' | grep -v '^Database')"
-SAVE_DBS="information_schema mysql performance_schema"
-SIZE_BEFORE=$(du /var/lib/mysql/ibdata1 | cut -d$'\t' -f1)
-for DB in $DBS; do
-  DROP=1
-  for SAVE_DB in $SAVE_DBS; do
-    if [ $SAVE_DB == $DB ]; then
-      DROP=0
-      break
-    fi
-  done
-  if [ $DROP == 1 ]; then
-    echo "Dumping $DB database..."
-    $MYSQL_DUMP -u$MYSQL_USER -p$MYSQL_PASS $DB | gzip > $DB.sql.gz
-    if [ -f $DB.sql.gz ]; then
-      echo "Database $DB backed up successfully."
-      echo "Dropping $DB database..."
-      $MYSQL -u$MYSQL_USER -p$MYSQL_PASS -e "drop database \`$DB\`"
-    else
-      echo "Backup of database $DB failed."
-      exit 1
-    fi
-  fi
+MYSQL_DATA_DIR=/var/lib/mysql
+
+
+# Functions
+prompt_yn() {	# params: message
+	msg="$*"
+	if [ -z "$msg" ]; then
+		msg='Are you sure?'
+	fi
+	result=-1
+	while true; do
+		read -rp "$msg [yes/no] " result
+		case $result in
+			[Yy]* ) result=0; break;;
+			[Nn]* ) result=1; break;;
+			*) echo "Please answer yes or no.";;
+		esac
+	done
+	return $result
+}
+warn() {	# params: message
+    echo "$@" >&2
+}
+die() {	# params: exit code; message
+    rc="$1"
+    shift
+    warn "$@"
+    exit "$rc"
+}
+usage() {
+	msg="$*"
+	this=$(basename "$0")
+	if [ -n "$msg" ]; then
+		warn "$this: ERROR: $msg"
+	fi
+	warn "Usage: $this [-dk]"
```

Comment:
`warn "Usage: $this [-cdhk]"`?

Reply:
This has been fixed/updated in my latest commit:
https://github.com/cmanley/shrink-ibdata1/commit/4804e224cf24a358e9945da76b9585b811e96c24